File: javascript/865.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
let t;
let d;
var traverse = function (rt, h) {
    if (rt == null)
        return h;
    let l = traverse(rt.left, h + 1);
    let r = traverse(rt.right, h + 1);
    if (l == r && l >= d) {
        t = rt;
        d = l;
    }
    return (l > r) ? l : r;
};
var subtreeWithAllDeepest = function (root) {
    d = 0;
    traverse(root, 0);
    return t;
};

File: python3/865.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.t = None
        self.d = 0

    def traverse(self, root, h):
        if not root:
            return h
        l = self.traverse(root.left, h + 1)
        r = self.traverse(root.right, h + 1)
        if l == r and l >= self.d:
            self.t = root
            self.d = l
        return max(l, r)

    def subtreeWithAllDeepest(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        self.traverse(root, 0)
        return self.t
